src/services/new/witness/index.ts

```diff
@@ -688,6 +688,87 @@ export class WitnessServiceV2 {
 
     }
 
+    async getWitnessActiveScore(block_height: number) {
+      const blockCount = 60
+      // const targetConstraints = {
+      //   // low: 7/10,
+      //   // high: 9/10
+      // }
+      const targetConstraints = [
+        [0, 0.6],
+        [0.75, 1],
+        [0.8, 1.3]
+      ]
+      
+      const lastXBlocks = await this.blockHeaders.find({
+        slot_height: {
+          $lte: block_height
+        }
+      }, {
+        sort: {
+          slot_height: -1
+        },
+        limit: blockCount
+      }).toArray()
+      
+      const expectedBlockScore: Record<string, number> = { 
+        
+      }
+      
+      const scoreMap: Record<string, number> = {
+        
+      }
+      
+      
+      for(let block of lastXBlocks) {
+        const {slot_height, signers} = block 
+        const election = await this.self.electionManager.getValidElectionOfblock(slot_height)
+  
+        const memberSet = election.members.map(e => e.account)
+        for(let member of memberSet) {
+          if(!scoreMap[member]) {
+            scoreMap[member] = 0
+          }
+          if(!expectedBlockScore[member]) {
+            expectedBlockScore[member] = 0
+          }
+          expectedBlockScore[member] = expectedBlockScore[member] + 1
+        }
+        
+        for(let account of signers) {
+          if(!scoreMap[account]) {
+            scoreMap[account] = 0
+          }
+          scoreMap[account] = scoreMap[account] + 1
+        }
+      }
+
+      let out = {}
+      for(let [key, value] of Object.entries(scoreMap)) {
+        const pctRaw = (value / expectedBlockScore[key])
+
+
+        let weight;
+        for(let tc of targetConstraints) {
+          const [t, w] = tc as any
+          if(pctRaw >= t) {
+            weight = w;
+          }
+        }
+        
+
+        out[key] = {
+          account: key,
+          pct: Math.round(pctRaw * 100),
```

Comment:
This is the pct of the blocks a node has signed, not of the entire network. This will always be less than or equal to 100. Total blocks in elections included by node / total signed blocks = pct

Reply:
I see. I was trying to point out the potential for floating-point rounding errors